app/javascript/controllers/distribution_confirmation_controller.js

```diff
@@ -0,0 +1,128 @@
+ * when the form data can pass initial validation.
+ * If the pre-check fails, it submits the form to the server for full validation and render with the errors.
+
+ * The modal shows the user what they're about to submit, including items and quantities to distribute.
+ * If multiple items have the same name, the quantities will be combined into a single row for display.
+ * If the user clicks the "Yes..." button from the modal, it submits the form.
+ * If the user clicks the "No..." button from the modal, it closes and user remains on the same url.
+ */
+export default class extends Controller {
+  static targets = [
+    "modal",
+    "form",
+    "partnerSelection",
+    "storageSelection",
+    "partnerName",
+    "storageName",
+    "itemSelection",
+    "quantity",
+    "tbody"
+  ]
+
+  static values = {
+    preCheckPath: String
+  }
+
+  openModal(event) {
+    event.preventDefault();
+    // this.debugFormData();
+
+    const formData = new FormData(this.formTarget);
+    const formObject = this.buildNestedObject(formData);
+
+    fetch(this.preCheckPathValue, {
+      method: "POST",
+      headers: {
+        "Content-Type": "application/json",
+        Accept: "application/json"
+      },
+      body: JSON.stringify(formObject),
+    })
+    .then((response) => response.json())
+    .then((data) => {
+      if (data.valid) {
+        console.log("=== DistributionConfirmationController VALID");
+        this.populatePartnerAndStorage();
+        this.populateItemsAndQuantities();
+        $(this.modalTarget).modal("show");
+      } else {
+        console.log("=== DistributionConfirmationController INVALID");
+        this.formTarget.requestSubmit();
+      }
+    })
+    .catch((error) => {
+      // Something went wrong trying to talk to server validation endpoint
+      // In this case, just submit the form as if the user had clicked Save.
+      // NICE TO HAVE: Send to bugsnag but need to install/configure https://www.npmjs.com/package/@bugsnag/js
+      console.log(`=== DistributionConfirmationController ERROR ${error}`);
+      this.formTarget.requestSubmit();
+    });
+  }
+
+  // Prepare the form data for submission as expected by Rails
+  buildNestedObject(formData) {
+    let formObject = {};
+
+    for (let [key, value] of formData.entries()) {
+      const keys = key.split(/[\[\]]+/).filter((k) => k); // Split and filter out empty strings
+      keys.reduce((obj, k, i) => {
+        if (i === keys.length - 1) {
+          obj[k] = value;
+        } else {
+          obj[k] = obj[k] || {};
+        }
+        return obj[k];
+      }, formObject);
+    }
+
+    return formObject;
+  }
+
+  populatePartnerAndStorage() {
+    const partnerName = this.partnerSelectionTarget.selectedOptions[0].text;
+    const storageName = this.storageSelectionTarget.selectedOptions[0].text;
+    this.partnerNameTarget.textContent = partnerName;
+    this.storageNameTarget.textContent = storageName;
```

Comment:
I'm not the biggest fan of basically reinventing ERB with this. :joy: The combination of line items is also something that we already have implemented in Ruby and have to do again here.

I'm wondering if we can reuse _just the ERB_ for this. Something like:

*   Send the fetch for validation
*   If valid, it sends the HTML to render, e.g. `{"valid": true, "body": "<div ...></div>"}`
*   Otherwise you submit the form as usual

You can build the distribution and line items and render the confirmation form with ERB that way. Something like

```ruby
def validate
  @dist = Distribution.new(distribution_params.merge(organization: current_organization))
  @dist.line_items.combine!
  if @dist.valid?
    body = render_to_string
    render json: {valid: true, body: body}
   else
    render json: {valid: false}
   end 
end
```

Then you just add `validate.html.erb` and use it like you would any other view. The Stimulus controller becomes super simple in that case, you just do something like `modalTarget.innerHTML = response.body`.

Reply:
Implemented, a few notes:

Had to specify the format as html (otherwise was trying to render json) and layout false, to get a focused view of only the modal body:
```ruby
body = render_to_string(template: 'distributions/validate', formats: [:html], layout: false)
```

Indeed this greatly simplified the stimulus controller, I was able to remove most of the targets. This could make it easier to re-use for other model confirmations in the future, by passing in a different `preCheckPath` value.